### ISSUE TITLE
Feasibility-aware promotion: land the validation-gate + final-selection feasibility rule from the CPAG branch as its own PR; headline tables report capture, violation, CVaR, worst, n together

### ISSUE THREAD

**Issue by GovitPowerz** [GovitPowerz/Aerocapture#109]
## Why (portfolio review 2026-09-16, P0 #2)

Promotion on `main` is val-RMS only. The validation gate promotes when `val_rms < best_val_cost` (`trainer.py:247-258`, `evaluate.py:163-222`), the islands `validate_each` does the same, and `final_select.select_final_individual` picks the lowest val RMS (`final_select.py:80,112-116`). Soft constraint penalties in the cost do not hold the feasibility line at the optimum: the paper's LSTM cell (`paper.typ:961`: capture 100.0, violation 15.6%) is the lowest-val-RMS run of its campaign and violates the integrated heat-load limit on 14-16% of evaluation draws; the per-draw fine-tune inherits it (`experiments/ou_marginal/RESULTS.md`: `ft_lstm_1082` 10.8% heat-load, `lstm_p1082_long` 17%). For an aerospace-control artifact this is a stronger reviewer objection than any architecture ranking: a leaderboard optimizer would deploy that policy; a requirements-first pipeline would not.

The rule to implement is lexicographic: (1) a candidate must satisfy every configured hard constraint below a stated violation ceiling on the validation pool; (2) among feasible candidates, minimize the performance metric; (3) the confirmatory pool is touched once, after selection. `IMPROVEMENTS.md:131-136` (9.14) already specifies this. It has never had a tracker issue, and #106 would fold it into TODO.md as prose.

**The implementation already exists.** Commit `0f6d531` on `feature/cpag-c1-rust-mvp` ("training: feasibility-aware validation gate (9.14) + CPAG C2 setup", 2026-07-16, six commits ahead of `main`) adds `constraint_violation_rates()` to `evaluate.py`, a `[optimizer] max_violation_rate` ceiling (default 0.0 = strict) in `optimizer.py`, the check in the validation gate, the islands path, `final_select` (fresh candidates evaluated via a new `evaluate_population_records_per_seed`; an infeasible fresh candidate never displaces the champion; the no-champion all-infeasible corner deploys best-RMS with a loud warning and `winner_feasible = False` in the sidecar), the pre-loop initial-champion validation, JSONL `violation_rates`, and `tests/test_feasibility_gate.py` (168 lines). The CPAG C2 campaign proved it live: the ungated pre-loop path anchored a 5.75% heat-load-violating gen-0 argmin as champion on the first launch; after the patch the gate rejected 21+ better-RMS infeasible candidates against 3 feasible promotions (`docs/plans/2026-07-19-cpag-c2-results.md` on the branch). The commit is entangled with CPAG (a 31k-line `cpag_confirmatory.json`, the CPAG training TOML, `train_all.sh`, `confirmatory_eval.py` changes), which is why it never reached `main`.

## Evidence (line anchors on `main` @ 55628eb)

- `src/python/aerocapture/training/trainer.py:247-258` (`promoted` -> `best_val_cost = gate.val_rms`), `:358-375` (final selection call), `:754-757` (islands champions).
- `src/python/aerocapture/training/evaluate.py:163-222` `run_validation_gate`: `promoted = val_rms < best_val_cost`.
- `src/python/aerocapture/training/final_select.py:76-130`: lowest val RMS over fresh U known; no constraint read.
- `src/python/aerocapture/training/report.py:427-495` `compute_eval_summary` already computes per-constraint exceedance percentages from the final-records matrix (the validation gate already feeds it for the TUI), so the rates exist; nothing acts on them.
- `articles/paper/scripts/confirmatory_eval.py:53` `REP_KEYS` carries capture and CVaR/max but no violation rate; `README.md:141-148` reports CVaR99.9 and ms/sim only; the paper table (`paper.typ:955-962`) has Capture % and Viol. % columns.
- Branch: `git show --stat 0f6d531` (15 files); `git diff main..feature/cpag-c1-rust-mvp --stat -- src/python/aerocapture/training tests/test_feasibility_gate.py tests/test_final_select.py`.

## Proposed shape (sketch, not prescriptive)

1. New branch `feature/feasibility-gate` from `main`; `git cherry-pick -n 0f6d531`, drop the CPAG-only hunks (`articles/paper/data/cpag_confirmatory.json`, `configs/training/msr_aller_cpag_train.toml`, `train_all.sh`, the CPAG lines in TODO/README/CLAUDE.md). Keep `confirmatory_eval.py --sim-timeout` / `--out` if they read as generic (they do: per-cell timeouts and own-pool output are protocol knobs); otherwise leave them to the CPAG decision. Expect conflicts in `trainer.py` / `train.py` / `final_select.py` (the trainer seam and the final-select CLI moved after 2026-07-16); resolve by intent, the branch's tests define it.
2. State the rule once: the `run_validation_gate` docstring, CONTEXT.md's **Champion** / **Final Selection** entries, and ADR-0006 "Feasibility before performance in champion selection" (decision, the LSTM evidence, the ceiling knob and its sample-size caveat: a strict 0 on a 400-sim pool lets one tail draw block every promotion, which is why CPAG C2 ran at 1%; the headline pools are 1000 sims where 0 is the right default).
3. Retro check of every deployed cell the README or paper quotes, via the `final_select` CLI (`python -m aerocapture.training.final_select <dir> --toml <config> --no-checkpoint-patch`) or a one-off script over `articles/paper/data/results.json` violation fields: list which cells pass at ceiling 0 on their validation pool. Expected: every per-draw scratch and the deployed `ft_mamba_962` pass (`RESULTS.md`: 0% violations); `ft_lstm_1082` and `lstm_p1082_long` fail. Record the list in the PR and mark infeasible cells as such wherever they are quoted.
4. Reporting: `confirmatory_eval.py` `REP_KEYS` gains per-constraint violation rates (heat flux, g-load, heat load) so the confirmatory JSON carries feasibility; the README results table and the paper's headline table report together: capture probability, constraint-violation probability, CVaR99.9 with its uncertainty, worst observed, n. Coordinate with #108 (which rewrites the same table) and #101 (RL row).
5. Resume semantics: a checkpoint written before the gate carries a champion that was never feasibility-checked. On resume, re-validate the champion with the gate (the `cost_transform`-changed path already re-validates; reuse it) and print a notice; an infeasible restored champion is kept as `last_validated` but its `best_val_cost` is reset to `inf` so the next feasible promotion replaces it (or: refuse to resume without `--accept-infeasible-champion`; maintainer call, default to the re-validate-and-reset).

## Constraints (load-bearing; read before touching)

- Numbers must not move for deployed cells: the gate changes which candidate promotes, not how any candidate scores. Goldens, `test_pyo3_matches_subprocess`, `tests/test_run_grid.py` unchanged. Bit-equivalence of the trainer seam (`experiments/trainer_seam_gate/`) is expected to break on runs where the gate rejects: rerun it once to confirm the ONLY divergence is a rejected promotion, and record that in the PR.
- Default ceiling 0.0 (strict). `[optimizer] max_violation_rate` is the one knob; `configs/training/common.toml` does not set it (the default applies).
- The `_SupportsPerSeedEval` / `_PerSeedEvaluator` Protocols gain `evaluate_population_records_per_seed`; the fakes in `tests/test_train_interrupt.py`, `tests/test_final_select.py`, `tests/test_island_model.py` must implement it.
- `constraint_violation_rates` reads limits from `cost_kwargs` (`read_cost_kwargs`, the canonical `[flight.constraints]` reader); no second reader.
- ADR-0002 (final selection on the validation pool, final eval report-only) is unchanged; the gate is a filter inside that rule.
- Feature branch, `smart-commit` at the end, `./lint_code.sh`, `./check_all.sh`, `uv run pytest tests -q -m "not slow"`, plus `uv run pytest tests/test_feasibility_gate.py tests/test_final_select.py tests/test_train_interrupt.py tests/test_island_model.py -q`.

## Acceptance

- [ ] `rg -n 'max_violation_rate' src/python/aerocapture/training/optimizer.py src/python/aerocapture/training/evaluate.py src/python/aerocapture/training/final_select.py src/python/aerocapture/training/trainer.py` -> hits in all four; `rg -n 'feasible' src/python/aerocapture/training/island_model.py` -> the islands path checks it.
- [ ] `uv run pytest tests/test_feasibility_gate.py -q` passes on `main`+PR; it contains a case where a lower-RMS infeasible candidate does NOT displace a feasible champion, a case for the pre-loop initial validation, and a case for the no-champion all-infeasible fallback (`winner_feasible == False`, warning printed).
- [ ] A validation JSONL record written by a smoke training run carries `violation_rates` and `feasible`; `final_selection.json` carries `winner_feasible`.
- [ ] `rg -n 'violation' articles/paper/scripts/confirmatory_eval.py` -> per-constraint rates in `REP_KEYS`; the README results table has capture, violation, CVaR99.9 +- se, max and n columns.
- [ ] `ls docs/adr/0006-*.md`; `rg -n 'feasib' CONTEXT.md` -> the Champion / Final Selection entries name the rule.
- [ ] The PR body lists every quoted deployed cell with its validation-pool violation rates at ceiling 0 and which ones fail.
- [ ] `git ls-files | rg cpag` -> empty on this branch (the CPAG artifacts did not ride along).
- [ ] Gates as listed in Constraints.

## Related

- `IMPROVEMENTS.md` 9.14 (the spec), `docs/plans/2026-07-19-cpag-c2-results.md` on `feature/cpag-c1-rust-mvp` (the live evidence), `TODO.md` CPAG Stage C2 ("requires the feasibility-aware validation gate").
- #106 (would fold 9.14 into TODO.md as prose; this issue replaces that line item), #101 (same headline table), #108 (same table; the deployed cell must pass this gate).
- Index: #115.

**Comment by GovitPowerz** [comment]
Landed on `feature/feasibility-gate` (commit `937f4b4`, not yet pushed). Acceptance status:

- `max_violation_rate` wired at every promotion site (optimizer / evaluate / final_select / trainer / island_model); islands `revalidate_each` on resume honours it too (an infeasible resumed champion keeps `last_validated` but its RMS no longer anchors `best_val_cost`).
- `tests/test_feasibility_gate.py`: lower-RMS infeasible candidate does not displace a feasible champion; pre-loop initial validation (feasible anchors, infeasible does not, `improved` flag follows); no-champion all-infeasible fallback (`winner_feasible == False`, warning). Plus a resume case in `tests/test_resume_enhancements.py`.
- Validation JSONL records carry `feasible` + `violation_rates` on both the single-algo and islands paths; `final_selection.json` carries `winner.feasible`.
- `confirmatory_eval.py`: per-constraint rates (heat flux, g-load, heat load, any) per replicate, aggregated with the tail stats. README results table now quotes capture %, violation %, CVaR99.9 +- se, max next to the tail number (n = 10^6 in the caption).
- ADR-0006 written; CONTEXT.md Validation Gate / Champion / Final Selection entries name the rule; CLAUDE.md, ARCHITECTURE.md, IMPROVEMENTS 9.14 pointer, TODO.md updated.
- No CPAG artifact rides along (`git ls-files | rg cpag` -> 0).
- Gates: `lint_code.sh`, `check_all.sh`, `pytest -m "not slow"` (1354 passed), islands + PSO end-to-end smokes, and the trainer-seam bit-equivalence gate. Seam-gate divergences from `main` are exactly rejected infeasible promotions on the 20-sim gate pool (gate a: heat flux on 2/20 draws for every validated candidate, so the run ends on the no-champion fallback; gate c: PSO island gen-0 heat load 1/20, rejected, all later records identical; gate b identical modulo the two new keys). Recorded in `experiments/trainer_seam_gate/README.md`.

Retro check, bundled final-eval pools (n = 1000, ceiling 0): 114 of 122 cells pass. Fail:

| cell | rate |
|---|---|
| headline/lstm_p1082 | heat load 15.6% |
| training_n_sims/adaptive_2 | heat load 0.4% |
| classical_baselines/piecewise_constant | heat flux 1.1% |
| classical_baselines/equilibrium_glide | heat flux 0.5% |
| legacy/neural_network_scaledpi_pso | heat flux 0.3% |
| classical_baselines/ftc | heat flux 0.2% |
| legacy/best_neural_network_joint | heat flux 0.1% |
| optimizer_dimensionality/ftc_pso | heat flux 0.1% |

Note for the maintainer: at strict 0, three classical baselines would have been blocked by 1-11 heat-flux draws per 1000. The default stays 0.0 (ADR-0006 states the sample-size caveat); a small ceiling on the 1000-sim validation pool is a config decision, not a code one.

**Comment by GovitPowerz** [comment]
Review pass on PR #116 landed as a second commit. Note the ADR is numbered **0005** (not 0006 as the first comment says): the sequence had a gap, and 0006 is now reserved for #108's per-draw default ADR. Also added: the retro CLI re-validates checkpoint champions before trusting them, and an infeasible resumed champion competes in the fallback on both paths.
